src/pegen/tokenizer.py

```diff
@@ -1,27 +1,121 @@
+    any token used as is in the grammar to be associated with a unique token type
+    listed in the subclass used by the parser.
+
+    Among the five tokens listed below, any token that is not used in the grammar
+    and hence never looked up by the parser can be omitted or given an arbitrary
+    value.
+
+    """
+
+    NAME: TokenType
+    NUMBER: TokenType
+    STRING: TokenType
+    OP: TokenType
+    TYPE_COMMENT: TokenType
+
+    def __init__(self) -> None:
+        ttype = type(self.NAME)
+        self.__dict__ |= {k: v for k, v in type(self).__dict__.items() if isinstance(v, ttype)}
+
+    @abstractmethod
+    def get_name_from_type(self, ttype: TokenType) -> str:
+        pass
+
+    @abstractmethod
+    def is_whitespace_token(self, tok: TokenInfo[TokenType]) -> bool:
+        pass
+
+    @abstractmethod
+    def is_relevant_token(
+        self, tok: TokenInfo[TokenType], seen_tokens: Sequence[TokenInfo[TokenType]]
+    ) -> bool:
+        """Is the token relevant to the parser."""
+        pass
+
+    def format_tok(self, tok: TokenInfo[TokenType]) -> str:
+        return f"{tok.start[0]}.{tok.start[1]}: {self.get_name_from_type(tok.type)}:{tok.string!r}"
+
+    def shorttok(self, tok: TokenInfo[TokenType]) -> str:
+        return "%-25.25s" % self.format_tok(tok)
+
+
+class PythonTokenTypes(TokenTypes[int]):
+    """Token types corresponding to the Python lexers."""
+
+    NAME = token.NAME
+    NUMBER = token.NUMBER
+    STRING = token.STRING
+    OP = token.OP
+    TYPE_COMMENT = token.TYPE_COMMENT
+    ENDMARKER = token.ENDMARKER
+    NEWLINE = token.NEWLINE
+    INDENT = token.INDENT
+    DEDENT = token.DEDENT
+
+    def get_name_from_type(self, ttype: int) -> str:
+        return token.tok_name[ttype]
+
+    def is_whitespace_token(self, tok: TokenInfo[int]) -> bool:
+        return tok.type == self.ENDMARKER or (tok.type >= self.NEWLINE and tok.type <= self.DEDENT)
+
+    def is_relevant_token(
+        self, tok: TokenInfo[TokenType], seen_tokens: Sequence[TokenInfo[TokenType]]
+    ) -> bool:
+        if tok.type in (token.NL, token.COMMENT):
+            return False
+        if tok.type == token.ERRORTOKEN and tok.string.isspace():
+            return False
+        if tok.type == self.NEWLINE and seen_tokens and seen_tokens[-1].type == self.NEWLINE:
+            return False
+        return True
+
+
+class Tokenizer(Generic[TokenType]):
+    """Caching wrapper for the token generator.
+
+    The peek method makes some assumptions on which tokens should be skipped
+    based on their types. On should overwrite it in a subclass if those
+    hypothesis do not match the parsed language.
 
-    This is pretty tied to Python's syntax.
     """
 
-    _tokens: List[tokenize.TokenInfo]
+    _tokens: List[TokenInfo[TokenType]]
 
     def __init__(
-        self, tokengen: Iterator[tokenize.TokenInfo], *, path: str = "", verbose: bool = False
+        self,
+        tokengen: Iterator[TokenInfo[TokenType]],
```

Comment:
why not name it `token_generator`?

Reply:
This predate this work so I did not feel like changing it.